Allow command-line arguments to be grouped in the usage.

nh13 opened the pull request:
@tfenne I am open to the formatting of the "group name" in the usage.  I have not implemented support for a description for each group; I am leaving that for latter if needed (I don't need it)

Commit message:

> The @arg annotation has a "group" member that can be set.  Leaving it
> blank (an empty string) will cause it to be part of the default group,
> and listed at the top of the usage.  The groups are then displayed in
> the usage in alphabetical order, with sections for required and optional
> command-line arguments for each group.

From the tests, this class:
```
@clp(description = "", group = classOf[TestGroup], hidden = true)
private case class MultipleGroupNames
(
  @arg(group="1") oneRequired: Int,
  @arg(group="1") oneOptional: Option[Int] = None,
  @arg(group="2") twoRequired: Int,
  @arg(group="3") threeOptional: Int = 1,
  @arg noGroup: Int = 3
)
```

gives this usage:

```
USAGE: MultipleGroupNames [arguments]
------------------------------------------------------------------------------------------------------------------------


MultipleGroupNames Optional Arguments:
------------------------------------------------------------------------------------------------------------------------
--no-group=Int                Undocumented option Default: 3.

MultipleGroupNames (1) Required Arguments:
------------------------------------------------------------------------------------------------------------------------
--one-required=Int            Undocumented option

MultipleGroupNames (1) Optional Arguments:
------------------------------------------------------------------------------------------------------------------------
--one-optional=Int            Undocumented option

MultipleGroupNames (2) Required Arguments:
------------------------------------------------------------------------------------------------------------------------
--two-required=Int            Undocumented option

MultipleGroupNames (3) Optional Arguments:
------------------------------------------------------------------------------------------------------------------------
--three-optional=Int          Undocumented option Default: 1.

nh13 committed:
Allow command-line arguments to be grouped in the usage.

The @arg annotation has a "group" member that can be set.  Leaving it
blank (an empty string) will cause it to be part of the default group,
and listed at the top of the usage.  The groups are then displayed in
the usage in alphabetical order, with sections for required and optional
command-line arguments for each group.

Includes a fix for missing square brackets when outputting the default
and true/false values.

Optional arguments with an "empty" default (ex. Option with a default of
None, or Seq with a default of Seq.empty), display "Optional" instead of
nothing at the end of the command line description.

nh13 commented on src/main/scala/com/fulcrumgenomics/sopt/cmdline/ClpReflectiveBuilder.scala:
```diff
@@ -154,6 +154,7 @@ private[sopt] class ClpArgument(declaringClass: Class[_],
   } else {
     throw new IllegalStateException("Calling maxElements on an argument that is not a collection.")
   }
+  lazy val groupName: Option[String] = annotation.map(_.group()).filter(_.nonEmpty)
```
remove me

tfenne commented on src/main/scala/com/fulcrumgenomics/sopt/cmdline/CommandLineProgramParser.scala:
```diff
@@ -227,19 +227,34 @@ class CommandLineProgramParser[T](val targetClass: Class[T], val includeSpecialA
+      .groupBy(_.groupName.getOrElse(""))
+
+    // Print out the commands by group
+    groups.keys.toSeq.sorted.map(_.capitalize).foreach { groupName =>
```
I think sorting here is undesirable.  We present options in the order they are declared in the CLP, giving the user control over things.  I think the thing to do here would be to instead iterate over all of the options and add the `group`s to a `LinkedHashSet`.  That way you'd get the unique set of groups, ordered by the first time they showed up in the options.  It's not perfect since you could still have:

```scala
@arg(group="foo") val x,
@arg(group="bar") val y,
@arg(group="foo") val z
```

but it would probably be the least surprising result and give the user some control over things.

tfenne commented on src/main/scala/com/fulcrumgenomics/sopt/cmdline/CommandLineProgramParser.scala:
```diff
@@ -227,19 +227,34 @@ class CommandLineProgramParser[T](val targetClass: Class[T], val includeSpecialA
+    // Print out the commands by group
+    groups.keys.toSeq.sorted.map(_.capitalize).foreach { groupName =>
+      val group: Seq[ClpArgument] = groups(groupName)
+      val (required, optional) = group.partition { !_.optional }
```
I also am beginning to wonder if the required/optional split is a good idea if there are multiple groups.  It leads to a lot of sections.  I'm not sure what else to suggest though that would make it clear what's optional and what's required.

tfenne reviewed (changes requested):
I think the ordering of the groups needs to be fixed, but if you agree then that's trivial to implement.

Also, I think the `Arg` api class in `Sopt` needs to be changed too to carry this forward, so that it can also be used in the documentation generation stuff.

nh13 commented:
@tfenne 

I have made the suggested changes.
In the current implementation, I have required args before optional args in a group, but lumped together in one section, and just have  different color for the optional arguments:
![screen shot 2017-10-03 at 11 40 32 am](https://user-images.githubusercontent.com/846274/31142849-52fa9aec-a830-11e7-8737-dabc474b2299.png)

I also fixed two bugs where square brackets were being stripped from the string displaying the default value, as well as the boolean true/false choices.

Finally, I updated the default value to say "Not Required" (see screenshot) when the underlying type is `Option[T]`, which I found was clearer than having just a description.

tfenne commented on src/main/java/com/fulcrumgenomics/sopt/cmdline/ArgAnnotation.java:
```diff
@@ -98,4 +98,11 @@
+     * command-line arguments will be displayed in the help in a separate section from those without a group name (the
+     * empty string).
+     */
+    String group() default "";
```
Nothing you can do, but it makes me sad we can't use things like Option here.

tfenne commented on src/main/scala/com/fulcrumgenomics/sopt/cmdline/ClpArgumentDefinitionPrinting.scala:
```diff
@@ -55,7 +57,8 @@ object ClpArgumentDefinitionPrinting {
       makeCollectionArity(argumentDefinition),
-      makeArgumentDescription(argumentDefinition, argumentLookup))
+      makeArgumentDescription(argumentDefinition, argumentLookup),
+      argumentDefinition.optional)
```
Could this be hoisted back onto the prior line please.

nh13 replied:
It can, but why should it?  It's not an argument to `makeArgumentDescription`

tfenne replied:
Nevermind, I misread.

tfenne commented on src/main/scala/com/fulcrumgenomics/sopt/cmdline/ClpArgumentDefinitionPrinting.scala:
```diff
@@ -93,24 +96,22 @@ object ClpArgumentDefinitionPrinting {
     val vs = defaultValuesAsSeq(value)
-    if (vs.isEmpty) "" else s"[Default: ${vs.mkString(", ")}]."
+    // NB: extra square brackets are inserted due to one set being stripped during markdown processing
+    KGRN(if (vs.isEmpty) "[[Not required]]." else s"[[$ArgumentDefaultValuePrefix ${vs.mkString(", ")}]].")
```
Two thoughts:

1. I'm not entirely sure that this catches all cases correctly?  Hopefully the tests will show me, but does it work correctly for e.g. required arguments with default values or optional values with None/Empty Seq etc.?
2. Could we change `Not required` to `Optional` please?

nh13 replied:
The method assumes that the argument is optional (`Seq`, `Option`, or has a default value).

1. There are tests in `ClpArgumentDefinitionPrintingTest`.
2. OK.

tfenne commented on src/main/scala/com/fulcrumgenomics/sopt/cmdline/CommandLineProgramParser.scala:
```diff
@@ -221,25 +240,47 @@ class CommandLineProgramParser[T](val targetClass: Class[T], val includeSpecialA
-        ClpArgumentDefinitionPrinting.printArgumentDefinitionUsage(builder, arg, argumentLookup)
+    // Get all the group names in the order they were first declared in a command line argument
+    val groupNames = mutable.LinkedHashSet[String]()
+    arguments.foreach { arg => groupNames.add(arg.groupName.getOrElse("").capitalize)}
```
How do you feel about removing `.capitalize`?  I tend to think that strings for display shouldn't get messed with.  Thought I guess this only toUppers the first char, so not so bad.

nh13 replied:
I want it to capitalize so I don't have to remember in my code.

tfenne reviewed (commented):
Looking good!  The `Sopt` API still needs updating, but this is 100% moving in the right direction.

One last thought: when there are defined groups and also some with group="", do you think we should give a default name to the group="" group so it doesn't look like an outlier?  E.g. "Other"?
